packages/block-library/src/cover/style.scss

```diff
@@ -30,19 +30,33 @@
 
 	&.has-background-dim::before {
 		content: "";
+		background-color: inherit;
+	}
+
+	&.has-background-dim:not(.has-background-gradient)::before,
+	.wp-block-cover__gradient-background {
 		position: absolute;
 		top: 0;
 		left: 0;
 		bottom: 0;
 		right: 0;
-		background-color: inherit;
-		opacity: 0.5;
 		z-index: z-index(".wp-block-cover.has-background-dim::before");
 	}
 
+	&.has-background-dim:not(.has-background-gradient)::before,
+	& .wp-block-cover__gradient-background {
+		opacity: 0.5;
+	}
+
+
 	@for $i from 1 through 10 {
-		&.has-background-dim.has-background-dim-#{ $i * 10 }::before {
-			opacity: $i * 0.1;
+		&.has-background-dim.has-background-dim-#{ $i * 10 } {
```

Comment:
Why not add it directly as a style attribute? Why limit the opacity to only 10 values. It would be good to allow anything between 0 and 100.